### PR TITLE
feat(primitives): `k256` crate fallback for `secp256k1` module

### PR DESCRIPTION
**Pull request by shekhirin**
Towards https://github.com/paradigmxyz/reth/issues/9478 for `reth-primitives` crate, inspired by https://github.com/paradigmxyz/reth/pull/9430

The `k256` implementation is cross-tested against `secp256k1`.

### CODE REVIEW

File: crates/primitives/Cargo.toml [hunk 1]
```diff
@@ -33,7 +33,8 @@ secp256k1 = { workspace = true, features = [
     "global-context",
     "recovery",
     "rand",
-] }
+], optional = true }
+k256 = { version = "0.13", default-features = false, features = ["ecdsa"], optional = true }
```
Comment (mattsse) [inline crates/primitives/Cargo.toml:37]: can we move this to workspace?

Comment (shekhirin) [reply crates/primitives/Cargo.toml]: done

File: Cargo.toml [hunk 2]
```diff
@@ -345,7 +345,7 @@ reth-optimism-rpc = { path = "crates/optimism/rpc" }
 reth-payload-builder = { path = "crates/payload/builder" }
 reth-payload-primitives = { path = "crates/payload/primitives" }
 reth-payload-validator = { path = "crates/payload/validator" }
-reth-primitives = { path = "crates/primitives", default-features = false, features = ["std"] }
+reth-primitives = { path = "crates/primitives", default-features = false, features = ["std", "secp256k1"] }
```
Comment (mattsse) [inline Cargo.toml:348]: we can probably remove "secp256k1" here seperately and enable it on the node level

File: crates/primitives/Cargo.toml [hunk 3]
```diff
@@ -33,7 +33,8 @@ secp256k1 = { workspace = true, features = [
     "global-context",
     "recovery",
     "rand",
-] }
+], optional = true }
+k256 = { workspace = true, default-features = false, features = ["ecdsa"] }
```
Comment (mattsse) [inline crates/primitives/Cargo.toml:37]: this is redundant
```suggestion
k256.workspace = true
```

File: crates/primitives/Cargo.toml [hunk 4]
```diff
@@ -81,10 +82,9 @@ pprof = { workspace = true, features = [
     "frame-pointer",
     "criterion",
 ] }
-secp256k1.workspace = true
 
 [features]
-default = ["c-kzg", "alloy-compat", "std", "reth-codec"]
+default = ["c-kzg", "alloy-compat", "std", "reth-codec", "secp256k1"]
```
Comment (mattsse) [inline crates/primitives/Cargo.toml:87]: we no longer enable c-kzg on workspace level, we need to be extra careful here and double check that we enable this somewhere.

I'm actually not sure where we enable revm c-kzg so we need to check and centralize these crucial settings, ideally on the node level

File: crates/primitives/src/transaction/util.rs [hunk 5]
```diff
@@ -47,17 +51,160 @@ pub(crate) mod secp256k1 {
...
+
+#[cfg(feature = "secp256k1")]
+pub(crate) mod secp256k1 {
+    pub use super::impl_secp256k1::*;
+}
+
+#[cfg(not(feature = "secp256k1"))]
+pub(crate) mod secp256k1 {
+    pub use super::impl_k256::*;
+}
```
Comment (mattsse) [inline crates/primitives/src/transaction/util.rs:115]: move above the modules

File: Cargo.toml (hunk 2 above)
Comment (mattsse) [inline Cargo.toml:348]: I would feel more comfortable if we don't do this in this pr and look at features separately

**Review by mattsse — CHANGES_REQUESTED** [review]
needs double check on c-kzg because this obfuscates revm features even more

https://github.com/paradigmxyz/reth/blob/0fe93a8de640770dc0e24ce4811dd0d1c0edbf04/crates/primitives/Cargo.toml#L103-L103

I think this will still get pulled in because we enable c-kzg for primitives somewhere else but still we need to do a proper workspace feature review

**Review by mattsse — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

File: crates/ethereum/evm/Cargo.toml [hunk 6]
```diff
@@ -15,7 +15,7 @@ workspace = true
 reth-chainspec.workspace = true
 reth-ethereum-forks.workspace = true
 reth-evm.workspace = true
-reth-primitives.workspace = true
+reth-primitives = { workspace = true, features = ["secp256k1"] }
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply crates/ethereum/evm/Cargo.toml]: oh okay, we need it only for tests anyway, so moved to dev deps

**Pull request merged**